trpcultivate_phenotypes/src/Plugin/Validators/ProjectGenusMatch.php

```diff
@@ -154,4 +177,107 @@ public function validateMetadata(array $form_values) {
+   *   to the user about the case that failed and the failed items from the
+   *   input file. Each item in the list contains either the project name or the
+   *   genus, whichever one caused the failure.
+   *
+   * @throws \Exception
+   *   - If the failure parameter was not formatted properly.
+   *   - If the case string returned by the validator implied validation passed.
+   *   - If the case string returned by the validator is not recognized.
+   */
+  public static function processSimpleList(array $failure, array $tokens = []) {
+
+    // @todo Re-add this check when the method is moved to its own Trait
+    // Check the format of the failure parameter.
+    // $this->checkValidationStatusArray($failure, 'ProjectGenusMatch');
+    // Combine our provided and our default token arrays. Because array_merge
+    // will overwrite values in the first array with values from the second
+    // array for the same keys, we provide our default tokens first.
+    $combined_tokens = array_merge(ProjectGenusMatch::$default_tokens, $tokens);
+
+    // Check for one of the expected cases. Use the message stored in the
+    // provided tokens array if set, otherwise use our default case message.
+    if ($failure['case'] == 'Project does not exist') {
+      $message = $combined_tokens['case-message-1'];
+      $items = [
+        '[project]: ' . $failure['failedItems']['project_provided'],
+      ];
+    }
+    elseif ($failure['case'] == 'Project has no genus set and could not compare with the genus provided') {
+      $message = $combined_tokens['case-message-2'];
+      $items = [
+        '[project]: ' . $failure['failedItems']['project_provided'],
+        'Genus: ' . $failure['failedItems']['genus_provided'],
+      ];
+    }
+    elseif ($failure['case'] == 'Genus does not match a genus set to the project') {
+      $message = $combined_tokens['case-message-3'];
+      $items = [
+        '[project]: ' . $failure['failedItems']['project_provided'],
+        'Genus: ' . $failure['failedItems']['genus_provided'],
+      ];
+    }
+    elseif ($failure['case'] == 'Project exists and project-genus match the genus provided') {
+      throw new \Exception('The case string returned by the ProjectGenusMatch validator implies validation passed, but valid is set to FALSE.');
+    }
+    else {
+      throw new \Exception('The case string returned by the ProjectGenusMatch validator is not recognized as a potential case.');
+    }
+
+    // Tripal Token Service.
+    $service_TripalTokensParser = \Drupal::service('tripal.token_parser');
+
+    // Now replace any tokens that are in our message.
+    $replaced_message = $service_TripalTokensParser->replaceTokens($message, $combined_tokens);
+    // Iterate through our items and replace any tokens there.
+    foreach ($items as $key => $item) {
+      $items[$key] = $service_TripalTokensParser->replaceTokens($item, $combined_tokens);
+    }
```

Comment:
For ease since we will be repeating this in every validator lets condense this and use [TripalTokenParser->replaceTokensArray()](https://github.com/tripal/tripal/blob/4.x/tripal/src/Services/TripalTokenParser.php#L61) to replace the items. Maybe we could also add a comment about why we're not using dependency injection here as a reminder to ourselves?

Something like:

```suggestion
    // Now replace any tokens that are in our message or items.
    // We use the Tripal Token Parser service to ensure we support more complicated tokens.
    // NOTE: we can't use depenendency injection since this is a static method.
    $service_TripalTokensParser = \Drupal::service('tripal.token_parser');
    $replaced_message = $service_TripalTokensParser->replaceTokens($message, $combined_tokens);
    $items = $service_TripalTokensParser->replaceTokensArray($items, $combined_tokens);
```

Reply:
Oooh, this is much simpler! Thanks for catching that. Addressed in commit https://github.com/TripalCultivate/TripalCultivate-Phenotypes/pull/159/commits/8ddf34b21be7278507f9e04e50cb75559966f4d5